### PR TITLE
Task-42687: Implement remove news article from news details

### PR DESCRIPTION
**Pull request by azayati**
Current behaviour:

- Only delete news article service and rest service is available
- When deleting the news article, the related activities are not removed
- It is not possible to remove a news article from news detail

New behaviour:

- When deleting the news article, the related activities are also removed
- It will be possible to remove a news article from news detail

### CODE REVIEW

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 1]
```diff
@@ -714,16 +745,76 @@ public Response deleteNews(@Context HttpServletRequest request,
...
+          if (newsToDeleteQueue.containsKey(id)) {
+            ExoContainerContext.setCurrentContainer(container);
+            RequestLifeCycle.begin(container);
+            try {
+              newsToDeleteQueue.remove(id);
+              newsService.deleteNews(id);
+            } catch (IllegalAccessException e) {
+              LOG.error("User '{}' attempts to delete a non authorized news", authenticatedUser, e);
+            } catch (Exception e) {
+              LOG.warn("Error deleting an news", e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
              LOG.warn("Error when deleting a news", e);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 2]
```diff
@@ -714,16 +745,76 @@ public Response deleteNews(@Context HttpServletRequest request,
...
+          @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error"), }
+  )
+  public Response undoDeleteNews(
+                                  @Context HttpServletRequest request,
+                                  @ApiParam(value = "News technical identifier", required = true)
+                                  @PathParam(
+                                    "newsId"
+                                  ) String newsId) {
+    if (StringUtils.isBlank(newsId)) {
+      return Response.status(Response.Status.BAD_REQUEST).entity("News identifier must be a positive integer").build();
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
      return Response.status(Response.Status.BAD_REQUEST).entity("News identifier must not be null or empty").build();
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 3]
```diff
@@ -714,16 +745,76 @@ public Response deleteNews(@Context HttpServletRequest request,
...
+        newsService.deleteNews(id);
+      }
       return Response.ok().build();
     } catch (Exception e) {
       LOG.error("Error when deleting the news with id " + id, e);
       return Response.serverError().build();
     }
   }
 
+  @Path("{newsId}/undoDelete")
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
  @Path("{id}/undoDelete")
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 4]
```diff
@@ -714,16 +745,76 @@ public Response deleteNews(@Context HttpServletRequest request,
...
+          @ApiResponse(code = HTTPStatus.BAD_REQUEST, message = "Invalid query input"),
+          @ApiResponse(code = HTTPStatus.FORBIDDEN, message = "Forbidden operation"),
+          @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "Unauthorized operation"),
+          @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error"), }
+  )
+  public Response undoDeleteNews(
+                                  @Context HttpServletRequest request,
+                                  @ApiParam(value = "News technical identifier", required = true)
+                                  @PathParam(
+                                    "newsId"
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
                                    "id"
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 5]
```diff
@@ -714,16 +745,76 @@ public Response deleteNews(@Context HttpServletRequest request,
...
+      value = {
+          @ApiResponse(code = HTTPStatus.NO_CONTENT, message = "Request fulfilled"),
+          @ApiResponse(code = HTTPStatus.BAD_REQUEST, message = "Invalid query input"),
+          @ApiResponse(code = HTTPStatus.FORBIDDEN, message = "Forbidden operation"),
+          @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "Unauthorized operation"),
+          @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error"), }
+  )
+  public Response undoDeleteNews(
+                                  @Context HttpServletRequest request,
+                                  @ApiParam(value = "News technical identifier", required = true)
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
                                  @ApiParam(value = "News node identifier", required = true)
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 6]
```diff
@@ -714,16 +745,76 @@ public Response deleteNews(@Context HttpServletRequest request,
...
+      String originalModifierUser = newsToDeleteQueue.get(newsId);
+      if (!originalModifierUser.equals(authenticatedUser)) {
+        LOG.warn("User {} attempts to cancel deletion of a news deleted by user {}", authenticatedUser, originalModifierUser);
+        return Response.status(Response.Status.FORBIDDEN).build();
+      }
+      newsToDeleteQueue.remove(newsId);
+      return Response.noContent().build();
+    } else {
+      return Response.status(Response.Status.BAD_REQUEST)
+                     .entity("News id was already deleted or isn't planned to be deleted")
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java]:
```suggestion
                     .entity("News with id {} was already deleted or isn't planned to be deleted", id)
```

File: services/src/main/resources/locale/portlet/news/News_en.properties [hunk 7]
```diff
@@ -185,4 +185,7 @@ news.permission.msg=You don't have the permission to write an article in this sp
 
 news.details.header.menu.share=Share
 news.details.header.menu.edit=Edit
+news.details.header.menu.delete=Delete
 
+news.undoRemoveActivity=Undo
```
Comment (azayati) [inline services/src/main/resources/locale/portlet/news/News_en.properties]:
```suggestion
news.details.undoDelete=Undo
```

File: services/src/main/resources/locale/portlet/news/News_en.properties [hunk 8]
```diff
@@ -185,4 +185,7 @@ news.permission.msg=You don't have the permission to write an article in this sp
...
+news.activityDeleteSuccess=Activity deleted
```
Comment (azayati) [inline services/src/main/resources/locale/portlet/news/News_en.properties]:
```suggestion
news.details.deleteSuccess=News deleted
```

File: services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java [hunk 9]
```diff
@@ -690,15 +716,20 @@ public Response viewNews(@Context HttpServletRequest request,
 
   @DELETE
   @Path("{id}")
-  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsRestResourcesV1.java:719]: Not needed IMHO, to keep just @Consumes

File: webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  computed: {
+    displayAlerts() {
+      return this.alerts && this.alerts.length;
+    },
+  },
+  created() {
+    this.$root.$on('news-notification-alert', alert => this.alerts.push(alert));
+    this.$root.$on('news-deleted', news => {
+      if (news && news.newsId) {
+        const clickMessage = this.$t('news.undoRemoveActivity');
```
Comment (azayati) [inline webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue]:
```suggestion
        const clickMessage = this.$t('news.details.undoDelete');
```

File: webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        const message = this.$t('news.activityDeleteSuccess');
```
Comment (azayati) [inline webapp/src/main/webapp/components/snackbar/ExoNewsNotificationAlerts.vue]:
```suggestion
        const message = this.$t('news.details.deleteSuccess');
```

File: webapp/src/main/webapp/news-details/components/ExoNewsDetails.vue [hunk 12]
```diff
@@ -227,10 +235,20 @@ export default {
         window.open('/', '_self');
       }
     },
+    reloadNews(newsId) {
```
Comment (azayati) [inline webapp/src/main/webapp/news-details/components/ExoNewsDetails.vue]: For what this function is needed ?

File: webapp/src/main/webapp/news-details/components/ExoNewsDetails.vue [hunk 13]
```diff
@@ -227,10 +235,20 @@ export default {
...
+      });
+    },
     editLink() {
       const editUrl = `${eXo.env.portal.context}/${eXo.env.portal.portalName}/news/editor?spaceId=${this.spaceId}&newsId=${this.news.newsId}&activityId=${this.activityId}`;
       window.open(editUrl, '_self');
     },
+    deleteNews() {
+      const deleteDelay = 10;
+      newsServices.deleteNews(this.newsId, deleteDelay)
+        .then(() => this.$root.$emit('news-deleted', this.news));
```
Comment (azayati) [inline webapp/src/main/webapp/news-details/components/ExoNewsDetails.vue]: you should redirect to homepage after delete success

File: webapp/src/main/webapp/services/newsServices.js [hunk 14]
```diff
@@ -259,3 +259,32 @@ export function canUserCreateNews(spaceId) {
...
+    localStorage.setItem('deletedNews', newsId);
+  }
+  return fetch(`${newsConstants.NEWS_API}/${newsId}?delay=${delay || 0}`, {
+    credentials: 'include',
+    method: 'DELETE'
+  }).then((resp) => {
+    if (resp && resp.ok) {
+      resp.json();
+    } else {
+      throw new Error('Error deleting news');
```
Comment (azayati) [inline webapp/src/main/webapp/services/newsServices.js]:
```suggestion
      throw new Error('Error when deleting news with id:' + newsId);
```

File: webapp/src/main/webapp/services/newsServices.js [hunk 15]
```diff
@@ -259,3 +259,32 @@ export function canUserCreateNews(spaceId) {
...
+
+export function undoDeleteNews(newsId) {
+  return fetch(`${newsConstants.NEWS_API}/${newsId}/undoDelete`, {
+    method: 'POST',
+    credentials: 'include',
+  }).then((resp) => {
+    if (resp && resp.ok) {
+      localStorage.removeItem('deletedNews');
+    } else {
+      throw new Error('Error undoing deleting news');
```
Comment (azayati) [inline webapp/src/main/webapp/services/newsServices.js]:
```suggestion
      throw new Error('Error when undoing deleting news with id:' + newsId);
```

File: webapp/src/main/webapp/news-details/components/ExoNewsDetailsActionMenu.vue [hunk 16]
```diff
@@ -50,6 +55,11 @@ export default {
       required: false,
       default: false
     },
+    showDeleteButton: {
```
Comment (azayati) [inline webapp/src/main/webapp/news-details/components/ExoNewsDetailsActionMenu.vue:58]: showDeleteButton should be false by default, true when the authenticated user is the news author (or space manager to be verified)

**Review by azayati — COMMENTED** [review]
@hassenelkamel-exo See review

File: webapp/src/main/webapp/services/newsServices.js (hunk 14 above)
Comment (hassenelkamel-exo) [reply webapp/src/main/webapp/services/newsServices.js]: with new Error you can only show new Error('String')

File: webapp/src/main/webapp/groovy/news/webui/activity/UINewsFullActivity.gtmpl [hunk 17]
```diff
@@ -191,8 +191,4 @@
   </div>
   </div><!--activityStream-->
   <% } // News archived %>
-  <% } else { // Activity deleted %>
-  <% uiform.begin() %>
-  <div class="activityStream deleted">$labelActivityHasBeenDeleted</div>
-  <% uiform.end() %>
-  <% } %>
+  <% }
```
Comment (hassenelkamel-exo) [inline webapp/src/main/webapp/groovy/news/webui/activity/UINewsFullActivity.gtmpl]: <% } %>

File: webapp/src/main/webapp/groovy/news/webui/activity/UINewsStreamActivity.gtmpl [hunk 18]
```diff
@@ -256,8 +256,4 @@
   </div> <!-- #boxContainer-->
   <% uiform.end() %>
 </div><!--activityStream-->
-<% } else { %> <!-- activity deleted -->
-<% uiform.begin() %>
-<div class="activityStream deleted">$labelActivityHasBeenDeleted</div>
-<% uiform.end() %>
-<% } %>
+<% }
```
Comment (hassenelkamel-exo) [inline webapp/src/main/webapp/groovy/news/webui/activity/UINewsStreamActivity.gtmpl]: <% } %>

File: webapp/src/main/webapp/groovy/news/webui/activity/UISharedNewsFullActivity.gtmpl [hunk 19]
```diff
@@ -229,8 +229,4 @@ if (activity && news) {
 </div><!--activityStream-->
 </div>
 <% } // News archived %>
-<% } else { // Activity deleted %>
-<% uiform.begin() %>
-<div class="activityStream deleted">$labelActivityHasBeenDeleted</div>
-<% uiform.end() %>
-<% } %>
+<% }
```
Comment (hassenelkamel-exo) [inline webapp/src/main/webapp/groovy/news/webui/activity/UISharedNewsFullActivity.gtmpl]: <% } %>

File: webapp/src/main/webapp/groovy/news/webui/activity/UISharedNewsStreamActivity.gtmpl [hunk 20]
```diff
@@ -242,8 +242,4 @@ def streamOwner = activity.getStreamOwner();
     </div> <!-- #boxContainer-->
     <% uiform.end() %>
 </div><!--activityStream-->
-<% } else { %> <!-- activity deleted -->
-<% uiform.begin() %>
-<div class="activityStream deleted">$labelActivityHasBeenDeleted</div>
-<% uiform.end() %>
-<% } %>
+<% }
```
Comment (hassenelkamel-exo) [inline webapp/src/main/webapp/groovy/news/webui/activity/UISharedNewsStreamActivity.gtmpl:245]: <% } %>

**Review by hassenelkamel-exo — CHANGES_REQUESTED** [review]

File: webapp/src/main/webapp/groovy/news/webui/activity/UISharedNewsStreamActivity.gtmpl (hunk 20 above)
Comment (hassenelkamel-exo) [inline webapp/src/main/webapp/groovy/news/webui/activity/UISharedNewsStreamActivity.gtmpl]: <% } %>

**Review by hassenelkamel-exo — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/exoplatform/news/NewsServiceImpl.java [hunk 21]
```diff
@@ -1230,6 +1250,17 @@ public void unarchiveNews(String newsId) throws Exception {
...
 
+  @Override
+  public boolean canDeleteNews(String posterId, String spaceId) {
+    org.exoplatform.services.security.Identity currentIdentity = getCurrentIdentity();
+    if (currentIdentity == null) {
+      return false;
+    }
+    String authenticatedUser = currentIdentity.getUserId();
+    Space currentSpace = spaceService.getSpaceById(spaceId);
+    return userACL.isSuperUser() || spaceService.isSuperManager(authenticatedUser) || spaceService.isManager(currentSpace, authenticatedUser);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/NewsServiceImpl.java]:
```suggestion
    return authenticatedUser.equals(posterId) || userACL.isSuperUser() || spaceService.isSuperManager(authenticatedUser) || spaceService.isManager(currentSpace, authenticatedUser);
```

File: services/src/main/java/org/exoplatform/news/model/News.java [hunk 22]
```diff
@@ -316,4 +318,12 @@ public String getAuthorAvatarUrl() {
   public void setAuthorAvatarUrl(String authorAvatarUrl) {
     this.authorAvatarUrl = authorAvatarUrl;
   }
+
+  public Boolean getCanDelete() {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/model/News.java]:
```suggestion
  public Boolean isCanDelete() {
```

File: webapp/src/main/webapp/news-details/main.js [hunk 23]
```diff
@@ -53,7 +54,8 @@ export function init(params) {
                   :activity-id="activityId"
                   :show-edit-button="showEditButton"
                   :show-pin-button="showPinButton"
-                  :show-share-button="showShareButton" />`,
+                  :show-share-button="showShareButton" 
+                  :show-delete-button="showEditButton"/>`,
```
Comment (azayati) [inline webapp/src/main/webapp/news-details/main.js]:
```suggestion
                  :show-delete-button="showDeleteButton"/>`,
```

File: webapp/src/main/webapp/news-details/components/ExoNewsDetails.vue [hunk 24]
```diff
@@ -127,6 +130,11 @@ export default {
       required: false,
       default: true
     },
+    showDeleteButton: {
+      type: Boolean,
+      required: false,
+      default: true
```
Comment (azayati) [inline webapp/src/main/webapp/news-details/components/ExoNewsDetails.vue]:
```suggestion
      default: false
```

**Review by azayati — COMMENTED** [review]
@hassenelkamel-exo see review
Some remarks from the previous review are missed, to be considered

**Review by hassenelkamel-exo — DISMISSED** [review]

**Review by azayati — COMMENTED** [review]
LGTM after QA validation

**Review by hassenelkamel-exo — DISMISSED** [review]

**Review by hassenelkamel-exo — APPROVED** [review]

**Pull request merged**